pkg/cmd/cli/cmd/rsync.go

```diff
@@ -0,0 +1,381 @@
+	}
+	glog.V(5).Infof("defaultLocalExecutor executing command: %s", strings.Join(command, " "))
+	cmd := exec.Command(command[0], command[1:]...)
+	cmd.Stdout = out
+	cmd.Stderr = errOut
+	cmd.Stdin = in
+	err := cmd.Run()
+	if err != nil {
+		glog.V(4).Infof("Error from local execution: %v", err)
+	}
+	return err
+}
+
+type defaultRemoteExecutor struct {
+	Namespace     string
+	PodName       string
+	ContainerName string
+	Client        *kclient.Client
+	Config        *kclient.Config
+}
+
+func (e *defaultRemoteExecutor) Execute(command []string, in io.Reader, out, errOut io.Writer) error {
+	glog.V(4).Infof("Remote executor running command: %s", strings.Join(command, " "))
+	execOptions := &kubecmd.ExecOptions{
+		In:            in,
+		Out:           out,
+		Err:           errOut,
+		Stdin:         in != nil,
+		Executor:      &kubecmd.DefaultRemoteExecutor{},
+		Client:        e.Client,
+		Config:        e.Config,
+		PodName:       e.PodName,
+		ContainerName: e.ContainerName,
+		Namespace:     e.Namespace,
+		Command:       command,
+	}
+	err := execOptions.Validate()
+	if err != nil {
+		glog.V(4).Infof("Error from remote command validation: %v", err)
+		return err
+	}
+	err = execOptions.Run()
+	if err != nil {
+		glog.V(4).Infof("Error from remote execution: %v", err)
+	}
+	return err
+}
+
+// NewCmdRsync creates a new sync command
+func NewCmdRsync(name, fullName string, f *clientcmd.Factory, out, errOut io.Writer) *cobra.Command {
+	o := RsyncOptions{
+		Out:           out,
+		Err:           errOut,
+		LocalExecutor: &defaultLocalExecutor{},
+		PipeExecutor:  &defaultPipeExecutor{},
+	}
+	cmd := &cobra.Command{
+		Use:     fmt.Sprintf("%s POD LOCAL_SOURCE REMOTE_DESTINATION", name),
+		Short:   "Copy local files to a pod",
+		Long:    rsyncLong,
+		Example: fmt.Sprintf(rsyncExample, fullName),
+		Run: func(c *cobra.Command, args []string) {
+			kcmdutil.CheckErr(o.Complete(f, c, args))
+			kcmdutil.CheckErr(o.Validate())
+			kcmdutil.CheckErr(o.Run())
+		},
+	}
+
+	cmd.Flags().StringVarP(&o.ContainerName, "container", "c", "", "Container within the pod")
+	cmd.Flags().StringVar(&o.RsyncExecutable, "rsync", "", "Path to rsync executable")
+	cmd.Flags().StringVar(&o.TarExecutable, "tar", "", "Path to tar executable")
+	cmd.Flags().BoolVarP(&o.Quiet, "quiet", "q", false, "Quiet copy")
+	return cmd
+}
+
+// Complete verifies command line arguments and loads data from the command environment
+func (o *RsyncOptions) Complete(f *clientcmd.Factory, cmd *cobra.Command, args []string) error {
+
+	if len(args) < 3 {
+		return kcmdutil.UsageError(cmd, "POD, LOCAL_SOURCE, and REMOTE_DESTINATION are required arguments")
+	}
+	if len(args) > 3 {
+
+	}
+	o.PodName = args[0]
+	o.Source = args[1]
+	o.Destination = args[2]
+
+	rsh := siblingCommand(cmd, "rsh")
+	rshCmd := []string{rsh, o.PodName}
+	if len(o.ContainerName) > 0 {
+		rshCmd = append(rshCmd, "-c", o.ContainerName)
+	}
+	rshCmd = append(rshCmd, "sh", "-c", "'exec \"$@\"'")
```

Comment:
The nested rsh command call you should be making is "oc rsh" - if you
generate the rsync command with podname:/foo instead of ignore:/foo, then
everything works correctly.  the exec @ thing is no longer necessary now
that rsh is fixed.

On Thu, Sep 24, 2015 at 11:18 AM, Cesar Wong <notifications@github.com>
wrote:

> In pkg/cmd/cli/cmd/rsync.go
> <https://github.com/openshift/origin/pull/4787#discussion_r40331177>:
>
> > +	if len(args) < 3 {
> > +		return kcmdutil.UsageError(cmd, "POD, LOCAL_SOURCE, and REMOTE_DESTINATION are required arguments")
> > +	}
> > +	if len(args) > 3 {
> > +
> > +	}
> > +	o.PodName = args[0]
> > +	o.Source = args[1]
> > +	o.Destination = args[2]
> > +
> > +	rsh := siblingCommand(cmd, "rsh")
> > +	rshCmd := []string{rsh, o.PodName}
> > +	if len(o.ContainerName) > 0 {
> > +		rshCmd = append(rshCmd, "-c", o.ContainerName)
> > +	}
> > +	rshCmd = append(rshCmd, "sh", "-c", "'exec \"$@\"'")
>
> I'm not using -i in the call to rsh, but are you talking about 'ignore:/'
> in the last parameter to rsync? If I don't have that there, rsync doesn't
> know that it's a remote destination.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4787/files#r40331177>.
>